src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1355 @@
+                OSMutexLock locker(this);
+                if (ipAcquiredSta_)
+                {
+                    ipAcquiredSta_ = false;
+                    mdns_disable_sta();
+                }
+            }
+            LOG(INFO, "wifi: STA lost IP.");
+            ip_acquired(IFACE_STA, false);
+            break;
+        }
+        case IP_EVENT_AP_STAIPASSIGNED:
+            ipLeased_ = true;
+            break;
+    }
+}
+
+//
+// EspIdfWiFi::init_config()
+//
+void EspIdfWiFi::init_config()
+{
+    esp_err_t result = nvs_open(NVS_NAMESPACE_NAME, NVS_READWRITE, &cfg_);
+    if (result != ESP_OK)
+    {
+        LOG(LEVEL_ERROR, "wifi: Error %s opening NVS handle.",
+            esp_err_to_name(result));
+        return;
+    }
+
+    // User configuration.
+    size_t len = sizeof(userCfg_);
+    result = nvs_get_blob(cfg_, NVS_KEY_USER_NAME, &userCfg_, &len);
+    switch(result)
+    {
+        case ESP_OK:
+            if (userCfg_.magic_ == WIFI_CONFIG_INIT_MAGIC &&
+                len == sizeof(userCfg_))
+            {
+                // Already initialized.
+                break;
+            }
+            // Not initialized yet.
+            // fall through
+        case ESP_ERR_NVS_NOT_FOUND:
+            // fall through
+        case ESP_ERR_NVS_INVALID_LENGTH:
+            // Initialize WiFi user configuration to factory defaults.
+            factory_default();
+            nvs_set_blob(cfg_, NVS_KEY_USER_NAME, &userCfg_, sizeof(userCfg_));
+            nvs_commit(cfg_);
+            break;
+        default:
+            LOG(LEVEL_ERROR, "wifi: Error %s getting userCfg_.",
+                esp_err_to_name(result));
+            break;
+    }
+
+    // Private configuration.
+    len = sizeof(privCfg_);
+    result = nvs_get_blob(cfg_, NVS_KEY_LAST_NAME, &privCfg_, &len);
+    switch (result)
+    {
+        case ESP_OK:
+            if (len == sizeof(privCfg_))
+            {
+                // Already initialized.
+                break;
+            }
+            // Not initialized yet.
+            // fall through
+        case ESP_ERR_NVS_NOT_FOUND:
+            // fall through
+        case ESP_ERR_NVS_INVALID_LENGTH:
+            // Reset to factory defaults.
+            memset(&privCfg_, 0, sizeof(privCfg_));
+            nvs_set_blob(cfg_, NVS_KEY_LAST_NAME, &userCfg_, sizeof(userCfg_));
+            nvs_commit(cfg_);
+            break;
+        default:
+            LOG(LEVEL_ERROR, "wifi: Error %s getting privCfg_.",
+                esp_err_to_name(result));
+            break;
+    }
+}
+
+//
+// EspIdfWiFi::init_wifi()
+//
+void EspIdfWiFi::init_wifi(WlanRole role)
+{
+    if (role == WlanRole::DEFAULT_ROLE)
+    {
+        role = userCfg_.mode_;
+    }
+
+    ESP_ERROR_CHECK(esp_netif_init());
+
+    /// @todo Do we need this?
+    esp_err_t result = esp_event_loop_create_default();
+    if (result != ESP_OK)
+    {
+        LOG(LEVEL_ERROR, "wifi: esp_event_loop_create_default() failed: %s",
```

Comment:
we should likely use `LOG_ERROR(...)` if we make similar calls elsewhere

Reply:
This is coming from OpenMRN shorthand logging formatting https://github.com/bakerstu/openmrn/blob/120ad4ea4b92cd1a461b82ec369fd510c8b8d798/src/utils/logging.h#L124.